src/librbd/api/Mirror.cc

```diff
@@ -3146,44 +3260,67 @@ int Mirror<I>::group_promote(IoCtx& group_ioctx, const char *group_name,
-          snap->state == cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
-        break;
-      }
-    }
 
-    if (snap == snaps.rend()) {
-      lderr(cct) << "failed to rollback, no mirror group snapshot available"
-        << dendl;
+    bool requires_orphan = false;
+    std::string rollback_group_snap_id = {};
+    if (!can_create_primary_group_snapshot(group_ioctx, &snaps, false,
+                                           true, &requires_orphan,
+                                           rollback_group_snap_id)) {
+      lderr(cct) << "cannot promote" << dendl;
       return -EINVAL;
-    }
-    // Check for group membership match
-    std::vector<cls::rbd::GroupImageSpec> rollback_images;
-    for (auto& it : snap->snaps) {
-      rollback_images.emplace_back(it.image_id, it.pool);
-    }
+    } else if (!rollback_group_snap_id.empty()) {
+      ldout(cct, 10) << "requires_orphan: " << requires_orphan
+                     << ", rollback_group_snap_id: " << rollback_group_snap_id
+                     << dendl;
+
+      auto snap =  snaps.rbegin();
+      for (; snap != snaps.rend(); ++snap) {
+        if (snap->id == rollback_group_snap_id) {
+          break;
+        }
+      }
 
-    if (rollback_images != current_images) {
-      lderr(cct) << "group snapshot membership does not match group membership"
-                 << dendl;
-      return -EINVAL;
-    }
+      if (snap == snaps.rend()) {
+        ldout(cct, 10) << "rollback snapshot not found" << dendl;
+        return -EINVAL;
+      }
 
-    librbd::NoOpProgressContext prog_ctx;
-    r = Group<I>::snap_rollback(group_ioctx,
-                                group_name, snap->name.c_str(), prog_ctx);
-    if (r < 0) {
-      lderr(cct) << "failed to rollback to group snapshot: " << snap->name
-                 << " :" << cpp_strerror(r) << dendl;
-      return r;
+      // Check for group membership match
+      std::vector<cls::rbd::GroupImageSpec> rollback_images;
+      for (auto& it : snap->snaps) {
+        rollback_images.emplace_back(it.image_id, it.pool);
+      }
+
+      if (rollback_images != current_images) {
+        lderr(cct) << "group snapshot membership does not match group membership"
+                   << dendl;
+        return -EINVAL;
+      }
+
+      r = create_orphan_group_snapshot(group_ioctx, group_id,
+                                       mirror_group.global_group_id);
+      if (r < 0 ) {
+        lderr(cct) << "orphan snap creation failed: "
+                   << cpp_strerror(r) << dendl;
+        return r;
+      }
+
+      librbd::NoOpProgressContext prog_ctx;
+      r = Group<I>::snap_rollback(group_ioctx,
+                                  group_name, snap->name.c_str(), prog_ctx);
+      if (r < 0) {
+        lderr(cct) << "failed to rollback to group snapshot: " << snap->name
+                   << " :" << cpp_strerror(r) << dendl;
+        return r;
+      }
+      ldout(cct, 5) << "successfully rollbacked to group snapshot: "
```

Comment:
"successfully rolled back to group snapshot"

Reply:
I would also recommend using the snap_id in the message as mirror snap names are long and confusing.